stingray/fourier.py

```diff
@@ -0,0 +1,1286 @@
+
+    return power * 2. / N / dt
+
+
+def normalize_leahy_from_variance(power, variance, N):
+    """Leahy+83 normalization, from the variance of the lc.
+
+    Examples
+    --------
+    >>> mean = var = 100000.
+    >>> N = 1000000
+    >>> lc = np.random.poisson(mean, N).astype(float)
+    >>> pds = np.abs(fft(lc))**2
+    >>> pdsnorm = normalize_leahy_from_variance(pds, var, lc.size)
+    >>> np.isclose(pdsnorm[0], 2 * np.sum(lc), rtol=0.01)
+    True
+    >>> np.isclose(pdsnorm[1:N//2].mean(), poisson_level(norm="leahy"), rtol=0.01)
+    True
+    """
+    return power * 2. / (variance * N)
+
+
+def normalize_leahy_poisson(power, Nph):
+    """Leahy+83 normalization, from the variance of the lc.
+
+    Examples
+    --------
+    >>> mean = var = 100000.
+    >>> N = 1000000
+    >>> lc = np.random.poisson(mean, N).astype(float)
+    >>> pds = np.abs(fft(lc))**2
+    >>> pdsnorm = normalize_leahy_poisson(pds, np.sum(lc))
+    >>> np.isclose(pdsnorm[0], 2 * np.sum(lc), rtol=0.01)
+    True
+    >>> np.isclose(pdsnorm[1:N//2].mean(), poisson_level(norm="leahy"), rtol=0.01)
+    True
+    """
+    return power * 2. / Nph
+
+
+def normalize_periodograms(unnorm_power, dt, N, mean, variance=None, norm="abs", power_type="all"):
+    """Wrapper around all the normalize_NORM methods.
+
+    Normalize the real part of the cross spectrum to Leahy, absolute rms^2,
+    fractional rms^2 normalization, or not at all.
+
+    Parameters
+    ----------
+    unnorm_power: numpy.ndarray
+        The unnormalized cross spectrum.
+
+    dt: float
+        The sampling time of the light curve
+
+    mean: float
+        The mean counts per bin of the light curve (if a cross spectrum, the geometrical
+        mean of the counts per bin in the two channels)
+
+    N: int
+        The number of bins in the light curve
+
+    Other parameters
+    ----------------
+    variance: float
+        The average variance of the measurements in light curve (if a cross spectrum,
+        the geometrical mean of the variances in the two channels). **NOT** the
+        variance of the light curve, but of each flux measurement (square of
+        light curve error bar)! Only relevant for the Leahy normalization of
+        non-Poissonian data.
+
+    norm : str
+        One of `'leahy'` (Leahy+83), `'frac'` (fractional rms), `'abs'`
+        (absolute rms)
+
+    power_type : str
+        One of `'real'` (real part), `'all'` (all complex powers), `'abs'`
+        (absolute value)
+
+    Returns
+    -------
+    power: numpy.nd.array
+        The normalized co-spectrum (real part of the cross spectrum). For
+        'none' normalization, imaginary part is returned as well.
+    """
+
+    if norm == "leahy" and variance is not None:
+        pds = normalize_leahy_from_variance(unnorm_power, variance, N)
+    elif norm == "leahy":
+        pds = normalize_leahy_poisson(unnorm_power, N * mean)
```

Comment:
Statistical point, because this seems to have tripped us up a couple of times in recent months: so if I've understood the code correctly, I think it's important to note here that `mean * N` is *not* the same as `Nphot`. This is because `Nphot` is a Poisson-distributed variable with a rate parameter equal to `mean * N`, because it's the sum of all the individual Poisson variables. This means that if you generate a large number of realizations, each of these realizations in practice have a different `Nphot`, which we're not accounting for here. I don't know whether this matters in practice. I suspect that for averaged spectra generated from a large number of segments, it might skew the statistical properties. Whether that'll be noticeable, I don't know.